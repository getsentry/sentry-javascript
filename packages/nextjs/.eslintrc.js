module.exports = {
  env: {
    browser: true,
    node: true,
  },
  parserOptions: {
    jsx: true,
  },
  extends: ['../../.eslintrc.js'],
  rules: {
    '@sentry-internal/sdk/no-optional-chaining': 'off',
    '@sentry-internal/sdk/no-nullish-coalescing': 'off',
  },
  overrides: [
    {
      files: ['scripts/**/*.ts'],
      parserOptions: {
        project: ['../../tsconfig.dev.json'],
      },
    },
  ],
};
