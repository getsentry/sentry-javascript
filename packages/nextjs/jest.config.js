const baseConfig = require('../../jest/jest.config.js');

module.exports = {
  ...baseConfig,
};
