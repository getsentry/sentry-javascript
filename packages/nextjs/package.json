{
  "name": "@sentry/nextjs",
  "version": "8.34.0",
  "description": "Official Sentry SDK for Next.js",
  "repository": "git://github.com/getsentry/sentry-javascript.git",
  "homepage": "https://github.com/getsentry/sentry-javascript/tree/master/packages/nextjs",
  "author": "Sentry",
  "license": "MIT",
  "engines": {
    "node": ">=14.18"
  },
  "main": "build/cjs/index.server.js",
  "module": "build/esm/index.server.js",
  "types": "build/types/index.types.d.ts",
  "files": [
    "/build"
  ],
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "types": "./build/types/index.types.d.ts",
      "edge": {
        "import": "./build/esm/edge/index.js",
        "require": "./build/cjs/edge/index.js",
        "default": "./build/esm/edge/index.js"
      },
      "edge-light": {
        "import": "./build/esm/edge/index.js",
        "require": "./build/cjs/edge/index.js",
        "default": "./build/esm/edge/index.js"
      },
      "worker": {
        "import": "./build/esm/edge/index.js",
        "require": "./build/cjs/edge/index.js",
        "default": "./build/esm/edge/index.js"
      },
      "workerd": {
        "import": "./build/esm/edge/index.js",
        "require": "./build/cjs/edge/index.js",
        "default": "./build/esm/edge/index.js"
      },
      "browser": {
        "import": "./build/esm/index.client.js",
        "require": "./build/cjs/index.client.js"
      },
      "node": "./build/cjs/index.server.js",
      "import": "./build/esm/index.server.js"
    },
    "./async-storage-shim": {
      "import": {
        "default": "./build/esm/config/templates/requestAsyncStorageShim.js"
      },
      "require": {
        "default": "./build/cjs/config/templates/requestAsyncStorageShim.js"
      }
    },
    "./import": {
      "import": {
        "default": "./build/import-hook.mjs"
      }
    },
    "./loader": {
      "import": {
        "default": "./build/loader-hook.mjs"
      }
    }
  },
  "typesVersions": {
    "<4.9": {
      "build/npm/types/index.d.ts": [
        "build/npm/types-ts3.8/index.d.ts"
      ]
    }
  },
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@opentelemetry/instrumentation-http": "0.53.0",
    "@opentelemetry/semantic-conventions": "^1.27.0",
    "@rollup/plugin-commonjs": "26.0.1",
    "@sentry-internal/browser-utils": "8.34.0",
    "@sentry/core": "8.34.0",
    "@sentry/node": "8.34.0",
    "@sentry/opentelemetry": "8.34.0",
    "@sentry/react": "8.34.0",
    "@sentry/types": "8.34.0",
    "@sentry/utils": "8.34.0",
    "@sentry/vercel-edge": "8.34.0",
    "@sentry/webpack-plugin": "2.22.3",
    "chalk": "3.0.0",
    "resolve": "1.22.8",
    "rollup": "3.29.5",
    "stacktrace-parser": "^0.1.10"
  },
  "devDependencies": {
    "@types/resolve": "1.20.3",
    "eslint-plugin-react": "^7.31.11",
    "next": "13.2.0"
  },
  "peerDependencies": {
    "next": "^13.2.0 || ^14.0 || ^15.0.0-rc.0",
    "webpack": ">=5.0.0"
  },
  "peerDependenciesMeta": {
    "webpack": {
      "optional": true
    }
  },
  "scripts": {
    "build": "run-p build:transpile build:types",
    "build:dev": "yarn build",
    "build:transpile": "ts-node scripts/buildRollup.ts",
    "build:types": "run-s build:types:core build:types:downlevel",
    "build:types:core": "tsc -p tsconfig.types.json",
    "build:types:downlevel": "yarn downlevel-dts build/types build/types-ts3.8 --to ts3.8",
    "build:watch": "run-p build:transpile:watch build:types:watch",
    "build:dev:watch": "yarn build:watch",
    "build:transpile:watch": "nodemon --ext ts --watch src scripts/buildRollup.ts",
    "build:types:watch": "tsc -p tsconfig.types.json --watch",
    "build:tarball": "npm pack",
    "circularDepCheck": "madge --circular src/index.client.ts && madge --circular src/edge/index.ts && madge --circular src/index.server.ts && madge --circular src/index.types.ts",
    "clean": "rimraf build coverage sentry-nextjs-*.tgz",
    "fix": "eslint . --format stylish --fix",
    "lint": "eslint . --format stylish",
    "test": "yarn test:unit",
    "test:all": "run-s test:unit",
    "test:unit": "jest",
    "test:types": "cd test/types && yarn test",
    "test:watch": "jest --watch",
    "vercel:branch": "source vercel/set-up-branch-for-test-app-use.sh",
    "vercel:project": "source vercel/make-project-use-current-branch.sh",
    "yalc:publish": "yalc publish --push --sig"
  },
  "volta": {
    "extends": "../../package.json"
  },
  "sideEffects": false
}
