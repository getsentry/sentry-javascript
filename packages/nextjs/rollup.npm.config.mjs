import { makeBaseNPMConfig, makeNPMConfigVariants, makeOtelLoaders } from '@sentry-internal/rollup-utils';

export default [
  ...makeNPMConfigVariants(
    makeBaseNPMConfig({
      // We need to include `instrumentServer.ts` separately because it's only conditionally required, and so rollup
      // doesn't automatically include it when calculating the module dependency tree.
      entrypoints: [
        'src/index.server.ts',
        'src/index.client.ts',
        'src/client/index.ts',
        'src/server/index.ts',
        'src/edge/index.ts',
        'src/config/index.ts',
      ],

      // prevent this internal nextjs code from ending up in our built package (this doesn't happen automatically because
      // the name doesn't match an SDK dependency)
      packageSpecificConfig: {
        external: ['next/router', 'next/constants', 'next/headers', 'stacktrace-parser'],
      },
    }),
  ),
  ...makeNPMConfigVariants(
    makeBaseNPMConfig({
      entrypoints: [
        'src/config/templates/apiWrapperTemplate.ts',
        'src/config/templates/middlewareWrapperTemplate.ts',
        'src/config/templates/pageWrapperTemplate.ts',
        'src/config/templates/requestAsyncStorageShim.ts',
        'src/config/templates/sentryInitWrapperTemplate.ts',
        'src/config/templates/serverComponentWrapperTemplate.ts',
        'src/config/templates/routeHandlerWrapperTemplate.ts',
      ],

      packageSpecificConfig: {
        output: {
          // Preserve the original file structure (i.e., so that everything is still relative to `src`)
          entryFileNames: 'config/templates/[name].js',

          // this is going to be add-on code, so it doesn't need the trappings of a full module (and in fact actively
          // shouldn't have them, lest they muck with the module to which we're adding it)
          sourcemap: false,
          esModule: false,

          // make it so Rollup calms down about the fact that we're combining default and named exports
          exports: 'named',
        },
        external: [
          '@sentry/nextjs',
          'next/dist/client/components/request-async-storage',
          '__SENTRY_CONFIG_IMPORT_PATH__',
          '__SENTRY_WRAPPING_TARGET_FILE__',
          '__SENTRY_NEXTJS_REQUEST_ASYNC_STORAGE_SHIM__',
        ],
      },
    }),
  ),
  ...makeNPMConfigVariants(
    makeBaseNPMConfig({
      entrypoints: ['src/config/loaders/index.ts'],

      packageSpecificConfig: {
        output: {
          // Preserve the original file structure (i.e., so that everything is still relative to `src`)
          entryFileNames: 'config/loaders/[name].js',

          // make it so Rollup calms down about the fact that we're combining default and named exports
          exports: 'named',
        },
        external: ['@rollup/plugin-commonjs', 'rollup'],
      },
    }),
  ),
  ...makeOtelLoaders('./build', 'sentry-node'),
];
