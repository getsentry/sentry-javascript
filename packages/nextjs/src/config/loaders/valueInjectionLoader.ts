// Rollup doesn't like if we put the directive regex as a literal (?). No idea why.
/* eslint-disable @sentry-internal/sdk/no-regexp-constructor */

import type { LoaderThis } from './types';

export type ValueInjectionLoaderOptions = {
  values: Record<string, unknown>;
};

// We need to be careful not to inject anything before any `"use strict";`s or "use client"s or really any other directive.
// As an additional complication directives may come after any number of comments.
// This regex is shamelessly stolen from: https://github.com/getsentry/sentry-javascript-bundler-plugins/blob/7f984482c73e4284e8b12a08dfedf23b5a82f0af/packages/bundler-plugin-core/src/index.ts#L535-L539
const SKIP_COMMENT_AND_DIRECTIVE_REGEX =
  // Note: CodeQL complains that this regex potentially has n^2 runtime. This likely won't affect realistic files.
  // biome-ignore lint/nursery/useRegexLiterals: No user input
  new RegExp('^(?:\\s*|/\\*(?:.|\\r|\\n)*?\\*/|//.*[\\n\\r])*(?:"[^"]*";?|\'[^\']*\';?)?');

/**
 * Set values on the global/window object at the start of a module.
 *
 * Options:
 *   - `values`: An object where the keys correspond to the keys of the global values to set and the values
 *        correspond to the values of the values on the global object. Values must be JSON serializable.
 */
export default function valueInjectionLoader(this: LoaderThis<ValueInjectionLoaderOptions>, userCode: string): string {
  // We know one or the other will be defined, depending on the version of webpack being used
  const { values } = 'getOptions' in this ? this.getOptions() : this.query;

  // We do not want to cache injected values across builds
  this.cacheable(false);

  // Not putting any newlines in the generated code will decrease the likelihood of sourcemaps breaking
  const injectedCode =
    // eslint-disable-next-line prefer-template
    ';' +
    Object.entries(values)
      .map(([key, value]) => `globalThis["${key}"] = ${JSON.stringify(value)};`)
      .join('');

  return userCode.replace(SKIP_COMMENT_AND_DIRECTIVE_REGEX, match => {
    return match + injectedCode;
  });
}
