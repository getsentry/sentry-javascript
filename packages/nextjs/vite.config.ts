import { defineConfig } from 'vitest/config';
import baseConfig from '../../vite/vite.config';

export default defineConfig({
  ...baseConfig,
  test: {
    ...baseConfig.test,
    environment: 'node',
  },
});
