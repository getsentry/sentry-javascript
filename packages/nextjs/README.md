<p align="center">
  <a href="https://sentry.io/?utm_source=github&utm_medium=logo" target="_blank">
    <img src="https://sentry-brand.storage.googleapis.com/sentry-wordmark-dark-280x84.png" alt="Sentry" width="280" height="84">
  </a>
</p>

# Official Sentry SDK for Next.js

[![npm version](https://img.shields.io/npm/v/@sentry/nextjs.svg)](https://www.npmjs.com/package/@sentry/nextjs)
[![npm dm](https://img.shields.io/npm/dm/@sentry/nextjs.svg)](https://www.npmjs.com/package/@sentry/nextjs)
[![npm dt](https://img.shields.io/npm/dt/@sentry/nextjs.svg)](https://www.npmjs.com/package/@sentry/nextjs)

> See the [Official Sentry Next.js SDK Docs](https://docs.sentry.io/platforms/javascript/guides/nextjs/) to get started.

## Compatibility

Currently, the minimum supported version of Next.js is `13.2.0`.

## Installation

To get started installing the SDK, use the Sentry Next.js Wizard by running the following command in your terminal or
read the [Getting Started Docs](https://docs.sentry.io/platforms/javascript/guides/nextjs/):

```sh
npx @sentry/wizard@latest -i nextjs
```

The wizard will prompt you to log in to Sentry. After the wizard setup is completed, the SDK will automatically capture
unhandled errors, and monitor performance.

## Custom Usage

To set context information or to send manual events, you can use `@sentry/nextjs` as follows:

```ts
import * as Sentry from '@sentry/nextjs';

// Set user information, as well as tags and further extras
Sentry.setTag('user_mode', 'admin');
Sentry.setUser({ id: '4711' });
Sentry.setContext('application_area', { location: 'checkout' });

// Add a breadcrumb for future events
Sentry.addBreadcrumb({
  message: '"Add to cart" clicked',
  // ...
});

// Capture exceptions or messages
Sentry.captureException(new Error('Oh no.'));
Sentry.captureMessage('Hello, world!');
```

## Links

- [Official SDK Docs](https://docs.sentry.io/platforms/javascript/guides/nextjs/)
- [Sentry.io](https://sentry.io/?utm_source=github&utm_medium=npm_nextjs)
- [Sentry Discord Server](https://discord.gg/Ww9hbqr)
- [Stack Overflow](https://stackoverflow.com/questions/tagged/sentry)
