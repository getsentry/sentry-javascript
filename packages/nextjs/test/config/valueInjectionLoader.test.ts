import type { LoaderThis } from '../../src/config/loaders/types';
import type { ValueInjectionLoaderOptions } from '../../src/config/loaders/valueInjectionLoader';
import valueInjectionLoader from '../../src/config/loaders/valueInjectionLoader';

const defaultLoaderThis = {
  addDependency: () => undefined,
  async: () => undefined,
  cacheable: () => undefined,
  callback: () => undefined,
};

const loaderThis = {
  ...defaultLoaderThis,
  resourcePath: './client.config.ts',
  getOptions() {
    return {
      values: {
        foo: 'bar',
      },
    };
  },
} satisfies LoaderThis<ValueInjectionLoaderOptions>;

describe('valueInjectionLoader', () => {
  it('should correctly insert values for basic config', () => {
    const userCode = `
      import * as Sentry from '@sentry/nextjs';
      Sentry.init();
    `;

    const result = valueInjectionLoader.call(loaderThis, userCode);

    expect(result).toMatchSnapshot();
    expect(result).toMatch(';globalThis["foo"] = "bar";');
  });

  it('should correctly insert values with directive', () => {
    const userCode = `
      "use client"
      import * as Sentry from '@sentry/nextjs';
      Sentry.init();
    `;

    const result = valueInjectionLoader.call(loaderThis, userCode);

    expect(result).toMatchSnapshot();
    expect(result).toMatch(';globalThis["foo"] = "bar";');
  });

  it('should correctly insert values with directive and semicolon', () => {
    const userCode = `
      "use client";
      import * as Sentry from '@sentry/nextjs';
      Sentry.init();
    `;

    const result = valueInjectionLoader.call(loaderThis, userCode);

    expect(result).toMatchSnapshot();
    expect(result).toMatch(';globalThis["foo"] = "bar";');
  });

  it('should correctly insert values with directive and inline comments', () => {
    const userCode = `
      // test
      "use client";
      import * as Sentry from '@sentry/nextjs';
      Sentry.init();
    `;

    const result = valueInjectionLoader.call(loaderThis, userCode);

    expect(result).toMatchSnapshot();
    expect(result).toMatch(';globalThis["foo"] = "bar";');
  });

  it('should correctly insert values with directive and block comments', () => {
    const userCode = `
      /* test */
      "use client";
      import * as Sentry from '@sentry/nextjs';
      Sentry.init();
    `;

    const result = valueInjectionLoader.call(loaderThis, userCode);

    expect(result).toMatchSnapshot();
    expect(result).toMatch(';globalThis["foo"] = "bar";');
  });

  it('should correctly insert values with directive and multiline block comments', () => {
    const userCode = `
      /*
        test
      */
      "use client";
      import * as Sentry from '@sentry/nextjs';
      Sentry.init();
    `;

    const result = valueInjectionLoader.call(loaderThis, userCode);

    expect(result).toMatchSnapshot();
    expect(result).toMatch(';globalThis["foo"] = "bar";');
  });

  it('should correctly insert values with directive and multiline block comments and a bunch of whitespace', () => {
    const userCode = `
      /*
        test
      */




      "use client";



      import * as Sentry from '@sentry/nextjs';
      Sentry.init();
    `;

    const result = valueInjectionLoader.call(loaderThis, userCode);

    expect(result).toMatchSnapshot();
    expect(result).toMatch(';globalThis["foo"] = "bar";');
  });

  it('should correctly insert values with a misplaced directive', () => {
    const userCode = `
      console.log('This will render the directive useless');
      "use client";



      import * as Sentry from '@sentry/nextjs';
      Sentry.init();
    `;

    const result = valueInjectionLoader.call(loaderThis, userCode);

    expect(result).toMatchSnapshot();
    expect(result).toMatch(';globalThis["foo"] = "bar";');
  });
});
