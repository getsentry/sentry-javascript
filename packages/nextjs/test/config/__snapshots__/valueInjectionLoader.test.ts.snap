// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`valueInjectionLoader should correctly insert values for basic config 1`] = `
"
      ;globalThis[\\"foo\\"] = \\"bar\\";import * as Sentry from '@sentry/nextjs';
      Sentry.init();
    "
`;

exports[`valueInjectionLoader should correctly insert values with a misplaced directive 1`] = `
"
      ;globalThis[\\"foo\\"] = \\"bar\\";console.log('This will render the directive useless');
      \\"use client\\";



      import * as Sentry from '@sentry/nextjs';
      Sentry.init();
    "
`;

exports[`valueInjectionLoader should correctly insert values with directive 1`] = `
"
      \\"use client\\";globalThis[\\"foo\\"] = \\"bar\\";
      import * as Sentry from '@sentry/nextjs';
      Sentry.init();
    "
`;

exports[`valueInjectionLoader should correctly insert values with directive and block comments 1`] = `
"
      /* test */
      \\"use client\\";;globalThis[\\"foo\\"] = \\"bar\\";
      import * as Sentry from '@sentry/nextjs';
      Sentry.init();
    "
`;

exports[`valueInjectionLoader should correctly insert values with directive and inline comments 1`] = `
"
      // test
      \\"use client\\";;globalThis[\\"foo\\"] = \\"bar\\";
      import * as Sentry from '@sentry/nextjs';
      Sentry.init();
    "
`;

exports[`valueInjectionLoader should correctly insert values with directive and multiline block comments 1`] = `
"
      /*
        test
      */
      \\"use client\\";;globalThis[\\"foo\\"] = \\"bar\\";
      import * as Sentry from '@sentry/nextjs';
      Sentry.init();
    "
`;

exports[`valueInjectionLoader should correctly insert values with directive and multiline block comments and a bunch of whitespace 1`] = `
"
      /*
        test
      */




      \\"use client\\";;globalThis[\\"foo\\"] = \\"bar\\";



      import * as Sentry from '@sentry/nextjs';
      Sentry.init();
    "
`;

exports[`valueInjectionLoader should correctly insert values with directive and semicolon 1`] = `
"
      \\"use client\\";;globalThis[\\"foo\\"] = \\"bar\\";
      import * as Sentry from '@sentry/nextjs';
      Sentry.init();
    "
`;
